### PR TITLE
add failing test for #121

### PR DESCRIPTION
**Pull request by iamstarkov**

### CODE REVIEW

**Comment by kof** [comment]
Fixed, are we good to go?

**Comment by iamstarkov** [comment]
i am concerned that right now tests are covering only SSR

**Comment by iamstarkov** [comment]
while error happened without it

**Comment by iamstarkov** [comment]
added tests for that

File: src/createHoc.js [hunk 1]
```diff
@@ -73,7 +73,12 @@ export default (stylesOrCreator, InnerComponent, options = {}) => {
     }
 
     get manager() {
-      return this.context[ns.sheetsManager] || manager
+      if (providerId && this.context[ns.providerId] !== providerId) {
+        manager = new SheetsManager()
+      }
+      providerId = this.context[ns.providerId]
+
+      return manager
```
Comment (iamstarkov) [inline src/createHoc.js]: can you describe in plain text what is happening here?

Comment (iamstarkov) [inline src/createHoc.js]: nvm

**Comment by kof** [comment]
> i am concerned that right now tests are covering only SSR

JssProvider was involved, it is most of the time used for SSR, also it has reproduced the issue, so I wouldn't know what to cover otherwise.

**Comment by iamstarkov** [comment]
@kof original issue was reproducible with `JssProvider` but without ssr, so i added tests for that

**Comment by kof** [comment]
It was reproducible with both, renderToString and render, so it doesn't matter. The nature of the issue was independent of environment. I removed those other tests because after fixing the issue covered by ssr test other tests were fixed too, so I though it doesn't make sense to have tests where it is not 100% clear why they are needed.

**Comment by iamstarkov** [comment]
my idea behind additional test is that issue is reproducible without SSR, so i thought it makes sense to have additional test wo/ ssr

File: src/JssProvider.js [hunk 2]
```diff
@@ -27,7 +26,7 @@ export default class JssProvider extends Component {
       [ns.sheetOptions]: {
         generateClassName: createGenerateClassName()
       },
-      [ns.sheetsManager]: new SheetsManager(),
```
Comment (oliviertassinari) [inline src/JssProvider.js]: I wish we have kept that feature for https://github.com/callemall/material-ui/pull/7461.

File: src/createHoc.js [hunk 3]
```diff
@@ -73,7 +73,12 @@ export default (stylesOrCreator, InnerComponent, options = {}) => {
     }
 
     get manager() {
-      return this.context[ns.sheetsManager] || manager
+      if (providerId && this.context[ns.providerId] !== providerId) {
+        manager = new SheetsManager()
```
Comment (oliviertassinari) [inline src/createHoc.js]: I think that it's breaking concurrency in SSR env.

Comment (iamstarkov) [inline src/createHoc.js]: what do you mean?

**Comment by iamstarkov** [comment]
@oliviertassinari i cant find you comment

> I wish we have kept that feature for https://github.com/callemall/material-ui/pull/7461

**Comment by oliviertassinari** [comment]
I have removed this comment. I have added that feature to the MuiThemeProvider instead of the JssProvider.

**Comment by iamstarkov** [comment]
i mean we fixed a bug, why would you rely on it?

File: src/createHoc.js (hunk 3 above)
Comment (oliviertassinari) [inline src/createHoc.js]: Let's say you are using https://formidable.com/blog/2017/introducing-rapscallion/, you are going to leak information between requests. That's bad for the UI and for security.

Comment (kof) [inline src/createHoc.js]: need to add a test for this

**Pull request merged**